exp/capacity-planning.md

```diff
@@ -0,0 +1,96 @@
+
+You should [run benchmarks](https://discourse.ubuntu.com/t/how-to-run-benchmarks/17770) to test your application performance and fine-tune the best node and application configuration. Also consider whether your containers use a hardware or software [video encoder](https://discourse.ubuntu.com/t/application-manifest/24197#video-encoder) for video encoding, and what frame rate and resolution they require.
+
+Another aspect is, of course, how many users you expect and how many containers will be running simultaneously. If you expect the usage to be rather consistent, you might not need to plan for huge peeks in load. On the other hand, you must also consider the impact if your cluster runs out of resources and it is not possible anymore to start more containers.
+
+## An example calculation
+
+Let's consider an application that uses the `g4.3` instance type with 4 vCPU cores, 3 GB of RAM, 3 GB of disk space and 1 GPU slot. The application is quite CPU-intensive, which means you should not over-commit resources too much. You expect an average of 100 containers running at the same time, with peeks up to 200.
+
+Without over-commitment, you would require the following resources to fulfil the average demand of 100 containers:
+
+- vCPU cores: `100 * 4 = 400`
+- RAM: `100 * 3 GB = 300 GB`
+- Disk space: `100 * 3 GB = 300 GB`
+- GPU slots: `100 * 1 = 100`
+
+With a CPU allocation rate of 2, you can bring the requirement of 400 vCPU cores down to 200 cores. With a CPU allocation rate of 4, the requirement would be reduced to 100. With a memory allocation rate of 2, you can bring the memory requirement down to 150 GB:
+
+- vCPU cores: `100 * 4 = 400` - with CPU allocation rate 2: `200`
+- RAM: `100 * 3 GB = 300 GB` - with memory allocation rate 2: `150 GB`
+- Disk space: `100 * 3 GB = 300 GB`
+- GPU slots: `100 * 1 = 100`
+
+The current calculation does not take into account that there might be peeks of up to 200 simultaneous containers. To cover all peeks, you would require the following resources:
+
+- vCPU cores: `200 * 4 = 800` - with CPU allocation rate 2: `400`
+- RAM: `200 * 3 GB = 600 GB` - with memory allocation rate 2: `300 GB`
+- Disk space: `200 * 3 GB = 600 GB`
+- GPU slots: `200 * 1 = 200`
+
+To avoid your cluster running out of resources even at peek loads, you must size it accordingly (or dynamically scale it up and down, see [About clustering](https://discourse.ubuntu.com/t/about-clustering/17765)). If the impact of not being able to provide additional containers at peek loads is rather low, you could compromise on the following factors:
+
+- Use a higher CPU and/or memory allocation rate, which might decrease performance at peek loads.
+- Configure your cluster nodes to use more GPU slots per GPU, which might decrease video quality.
+- Tweak the resource specification for your application to give less resources to each container. The impact of doing this depends very much on your application.
```

Comment:
Tweak the resource specification for your application
->
Tweak the instance-type or resource specification for your application